gateway-server/src/main/java/org/apache/knox/gateway/websockets/JWTValidator.java

```diff
@@ -0,0 +1,288 @@
+            try {
+                publicKey = CertificateUtils.parseRSAPublicKey(verificationPEM);
+            } catch (ServletException e){
+                throw new RuntimeException("Failed to obtain public key: "+e);
+            }
+        }
+
+        //provider-level configuration
+        providerParamValue = params.get(TokenStateService.CONFIG_SERVER_MANAGED);
+
+        cookieName = params.get(KNOXSSO_COOKIE_NAME);
+        if (cookieName == null) {
+            cookieName = DEFAULT_SSO_COOKIE_NAME;
+        }
+        expectedIssuer =  params.get(JWT_EXPECTED_ISSUER);
+        if (expectedIssuer == null) {
+            expectedIssuer = JWT_DEFAULT_ISSUER;
+        }
+        expectedSigAlg =  params.get(JWT_EXPECTED_SIGALG);
+        if (expectedSigAlg == null) {
+            expectedSigAlg = JWT_DEFAULT_SIGALG;
+        }
+        authorityService = gatewayServices.getService(ServiceType.TOKEN_SERVICE);
+        if (isServerManagedTokenStateEnabled()) {
+            tokenStateService = gatewayServices.getService(ServiceType.TOKEN_STATE_SERVICE);
+        }
+        // Setup the verified tokens cache
+        signatureVerificationCache = SignatureVerificationCache.getInstance(
+                "knoxsso", new WebSocketFilterConfig(params));
+    }
+
+    private void extractToken(ServletUpgradeRequest req){
+        List<HttpCookie> ssoCookies = req.getCookies();
+        if (ssoCookies != null){
+            for (HttpCookie ssoCookie : ssoCookies) {
+                if (cookieName.equals(ssoCookie.getName())) {
+                    try {
+                        token = new JWTToken(ssoCookie.getValue());
+                        displayableTokenId = Tokens.getTokenIDDisplayText(TokenUtils.getTokenId(token));
+                        displayableToken = Tokens.getTokenDisplayText(token.toString());
+                        websocketLog.debugLog("found token:"+displayableToken+" id:"+displayableTokenId);
+                        return;
+                    } catch (ParseException e) {
+                        // Fall through to keep checking if there are more cookies
+                    }
+                }
+            }
+        }
+        log.missingBearerToken();
+        throw new RuntimeException("no Valid JWT found");
+    }
+
+    public JWT getToken(){
+        return token;
+    }
+
+    public String getUsername(){
+        return token.getSubject();
+    }
+
+    public boolean validate() {
+        // confirm that issuer matches the intended target
+        if (expectedIssuer.equals(token.getIssuer())) {
+            // if there is no expiration data then the lifecycle is tied entirely to
+            // the cookie validity - otherwise ensure that the current time is before
+            // the designated expiration time
+            try {
+                if (tokenIsStillValid()) {
+                    Date nbf = token.getNotBeforeDate();
+                    if (nbf == null || new Date().after(nbf)) {
```

Comment:
Why are we letting tokens with `null` nbf through? do we do this elsewhere?

Reply:
yes, i copied logic from AbstractJWTFilter.java  validateToken function line 335